jetty-util/src/main/java/org/eclipse/jetty/util/resource/ByteBufferResource.java

```diff
@@ -0,0 +1,230 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2019 Mort Bay Consulting Pty. Ltd.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.util.resource;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.charset.Charset;
+
+import org.eclipse.jetty.util.BufferUtil;
+
+/**
+ * A Memory only Resource with static content.
+ */
+public class ByteBufferResource extends Resource
```

Comment:
It's not a `ByteBuffer` resource, more a `Bytes` resource.
How those bytes are fed to the resource is an implementation detail, but `ByteBufferResource` seems to imply you have to use `java.nio.ByteBuffer` which is not the case (as you can pass in `String`, `byte[]`, etc.).
I like `BytesResource` better as it's more generic.

Reply:
I personally don't like `ByteResource` or `BytesResource` as that doesn't indicate that it's a memory only resource.

Lets take a look at the existing Resources (which all have URI references and behaviors) ...

* Resource - base class, represents an abstract resource
* URLResource - Resources representing a URL location - aka `http://` or `https://` (or any registered URL Stream Handler)
* PathResource - Resources representing a Path location - aka `file://`
* JarResource - Resources representing a JAR File. - aka `jar://`
* JarFileResource - Resources representing a File in a JAR File. - aka `jar:file://`
* BadResource - Resources representing a Bad Resource - aka `bad://` (really any scheme not supported above)
* EmptyResource - Resources representing an Empty Resource (no content) - aka `null`

And the one I'm proposing

* MemoryResource - Resources representing a wholly in memory resource (no directories support) - `memory://`

The reason I don't like "ByteResource" or "BytesResource" is that it doesn't represent anything specific. and the required URI is weird too.
That would mean we have `byte://` or `bytes://` which means what?

We have MemoryInputStream already in the (jetty-proxy) codebase, and it represents a InputStream that's fed entirely from memory.
Why can't we have MemoryResource as it's a Resource represented entirely from memory.
Using your own argument, "Byte" is an implementation detail, let it be "Memory" instead, and URI scheme `memory://`

Reply:
You don't pass this Resource a "memory".
For all other Resource implementations, you pass them a `URL`, a `Path`, a `JarFile`, etc. they are named after what they take at construction.
Here you pass bytes to this resource, hence the name I propose.

Reply:
I think `ByteResource` is the correct name, like `ByteChannel`.  It can be constructed with a string, and array or a buffer, but all provide bytes.

Reply:
Nah, still doesn't work for me.
After all ...
You don't pass "JarResource" a java.util.jar.JarFile or java.io.File or java.nio.files.Path, you pass it a URL, which either exists and has content, or not.
You don't pass "JarFileResource" a "JarFile" either, you pass it a URL.
You don't pass "BadResource" a "Bad" you give it a URL.

Point is, the name of the Resource isn't about what you give it, it's about what it represents (and just as important, it's URI scheme).

ByteResource could just as well be a stream of bytes off disk, or a network, or any of a dozen other things that have nothing to do with "statically declared, from memory only".
The only way you know how ByteResource behaves is from the limited set of constructors it has.
What if someone comes along and adds "public ByteResource(SocketChannel channel)"?

I started out with StaticResource, but that didn't sound right either.